clippy_lints/src/explicit_struct_update.rs

```diff
@@ -0,0 +1,147 @@
+    ///     a: 5,
+    ///     ..my_foo
+    /// };
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub EXPLICIT_STRUCT_UPDATE,
+    complexity,
+    "explicit struct updates in struct instantiations"
+}
+declare_lint_pass!(ExplicitStructUpdate => [EXPLICIT_STRUCT_UPDATE]);
+
+impl<'tcx> LateLintPass<'tcx> for ExplicitStructUpdate {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        if hir::is_range_literal(expr) {
+            // range literals are lowered to structs, false positive
+            return;
+        }
+
+        let ExprKind::Struct(path, fields, StructTailExpr::None) = expr.kind else {
+            return;
+        };
+
+        // the type of the struct
+        let ty = cx.typeck_results().expr_ty(expr);
+
+        // collect the fields that are being initialized with the same field from another struct of the same
+        // type
+        let update_fields = fields.iter().try_fold(
+            Vec::new(),
+            |mut acc: Vec<(&rustc_hir::Expr<'_>, &rustc_hir::Expr<'_>)>, f| {
+                if let ExprKind::Field(base_expr, field_ident) = f.expr.kind {
+                    if let Some(last) = acc.last() {
+                        match (last.1.kind, base_expr.kind) {
+                            (
+                                ExprKind::Path(hir::QPath::Resolved(_, hir::Path { res: res_a, .. })),
+                                ExprKind::Path(hir::QPath::Resolved(_, hir::Path { res: res_b, .. })),
+                            ) if res_a != res_b => return None, /* if we detect instantiation from multiple bases, we */
+                            // don't want to lint
+                            _ => (),
+                        }
+                    }
+
+                    if cx.typeck_results().expr_ty(base_expr) == ty && f.ident == field_ident {
+                        // accumulate the expressions mapping to the actual field expression, and the expression of the
+                        // base struct, we do this so we can determine if the base struct is the same for all
+                        acc.push((f.expr, base_expr));
+                    }
+                }
+
+                Some(acc)
+            },
+        );
+
+        let (update_base, update_fields): (_, Vec<_>) = match update_fields {
+            // we only care about the field expressions at this point
+            Some(fields) if fields.len() > 1 => (fields[0].1, fields.iter().map(|x| x.0.hir_id).collect()),
+            // no lint if there's no fields or multiple bases
+            _ => return,
+        };
+
+        // the field assignments we are keeping
+        let non_update_fields_spans: Vec<_> = fields
+            .iter()
+            .filter_map(|f| {
+                if update_fields.contains(&f.expr.hir_id) {
+                    None
+                } else {
+                    Some(f.span)
+                }
+            })
+            .collect();
```

Comment:
I think you could do this more concisely, with something like:

```suggestion
        // collect the fields that are being initialized with the same field from another struct of the same
        // type
        let (mut update_base, mut update_fields, mut non_update_fields_spans) = (None, vec![], vec![]);
        for field in fields {
            if let ExprKind::Field(base_expr, field_ident) = field.expr.kind
                && let ExprKind::Path(hir::QPath::Resolved(_, hir::Path { res, .. })) = base_expr.kind
                && field_ident == field.ident
                && cx.typeck_results().expr_ty(base_expr) == ty
            {
                if matches!(update_base.replace((res, base_expr.span)), Some((base_res, _)) if base_res != res) {
                    // Initialization from multiple bases
                    return;
                }
                update_fields.push(field.hir_id);
            } else {
                non_update_fields_spans.push(field.span);
            }
        }

        let update_base_span = match update_base {
            Some((_, span)) if update_fields.len() > 1 => span,
            _ => return,
        };
```